Added support of Smepmp in Sail RISCV

HamzaKh01 opened the pull request:
### Description
This PR includes the support of ePMP in sail risc-v. Model is verified by running the tests of ePMP (can be found [here](https://github.com/riscv-non-isa/riscv-arch-test/pull/301) ) on the RiscOf environment  which runs the tests on spike and sail and compare their signatures. So, it can be said that, this Sail model has a same behaviour for ePMP as of spike. Kindly review this PR. In addition the testplan for writting the tests is attached [here](https://docs.google.com/spreadsheets/d/1dDeSlpOrtMDFhkv7ysAbCgCjOKVzhhvxCY83DirL92w/edit#gid=0)

jrtc27 commented on .gitignore:
```diff
@@ -1,4 +1,7 @@
 /generated_definitions
+z3_problems
+c_emulator/riscv_sim_RV32
+c_emulator/riscv_sim_RV64
```
These are unrelated to the PR

HamzaKh01 replied:
Removed

jrtc27 commented on model/riscv_pmp_control.sail:
```diff
@@ -104,15 +119,52 @@ function pmpCheckRWX(ent, acc) = {
+  else 
+      match priv {
+          Machine => if   pmpLockBit(ent)
                then pmpCheckRWX(ent, acc)
```
Formatting is broken here

HamzaKh01 replied:
Formatting is fixed

jrtc27 commented on model/riscv_pmp_control.sail:
```diff
@@ -92,6 +92,21 @@ function pmpAddrRange(cfg: Pmpcfg_ent, pmpaddr: xlenbits, prev_pmpaddr: xlenbits
+
+val MMLcheck: (AccessType(ext_access_type)) -> bool effect {rreg}
+function MMLcheck(acc) = {
+  if ( mseccfg.MML() == 0b1)
+	then { 
+		    match acc {
+				Execute(_)  => false,
+      		   _       => true
+			  }
+			 }
+	else
+	     true	 
```
Formatting is unreadable, let alone conforming

HamzaKh01 replied:
Formatting is fixed, readability is improved

jrtc27 commented on model/riscv_pmp_control.sail:
```diff
@@ -92,6 +92,21 @@ function pmpAddrRange(cfg: Pmpcfg_ent, pmpaddr: xlenbits, prev_pmpaddr: xlenbits
   }
 }
 
+/* Check mseccfg.MMl=1 and mseccfg.MMWP=0 and make execute false for Mmode for matching region not found*/
```
This sentence doesn't make much sense, but don't just write out a copy of the implementation in English, document what it _means_

jrtc27 replied:
Also you need a space before `*/`

HamzaKh01 replied:
All comments are improved and formatting is done

jrtc27 commented on model/riscv_pmp_control.sail:
```diff
@@ -104,15 +119,52 @@ function pmpCheckRWX(ent, acc) = {
 
-// this needs to be called with the effective current privilege.
-val pmpCheckPerms: (Pmpcfg_ent, AccessType(ext_access_type), Privilege) -> bool
+/* this needs to be called with the effective current privilege.*/
```
Why? Though if you do this, at least get the formatting right and have a space before `*/`

HamzaKh01 replied:
Done

jrtc27 commented on model/riscv_pmp_control.sail:
```diff
@@ -104,15 +119,52 @@ function pmpCheckRWX(ent, acc) = {
+  if mseccfg.MML() == 0b1
+  then
+  {
+      if (ent.R() == 0b0 & ent.W() == 0b1)
```
Don't put parentheses when you don't need them, if x doesn't need any

HamzaKh01 replied:
Removed unnecessary parentheses

jrtc27 commented on model/riscv_pmp_control.sail:
```diff
@@ -218,7 +270,9 @@ function pmpCheck forall 'n, 'n > 0. (addr: xlenbits, width: atom('n), acc: Acce
     PMP_Fail     => false,
     PMP_Continue => match priv {
-                      Machine => true,
+                      Machine => if mseccfg.MMWP() == 0b1   /*If mseccfg.MMWP is set then default pmp policy for matching region not found is denied*/
```
Why not move this check into MMLcheck given this is the only place the function is called?

HamzaKh01 replied:
Yes, you are right, the function is removed and the check is moved where the function was called

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -100,6 +100,14 @@ bitfield Pmpcfg_ent : bits(8) = {
 }
 
+bitfield Mseccfg_ent : xlenbits = {
+  RLB : 2,        /* Rule Locking Bypass*/
+  MMWP : 1,       /* Machine Mode Whitelist Policy*/
```
```suggestion
  RLB : 2,        /* Rule Locking Bypass */
  MMWP : 1,       /* Machine Mode Whitelist Policy */
```

HamzaKh01 replied:
Fixed

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +162,28 @@ function pmpReadCfgReg(n) = {
+  
+/*Function implements 4b of Smepmp spec*/
+function checkExe(cfg: Pmpcfg_ent,v: bits(8)) -> Pmpcfg_ent =
+	 if ((mseccfg.MML() == 0b1) & (mseccfg.RLB() == 0b0) & (v[7] == bitone) & (v[2] == bitone) | (mseccfg.MML() == 0b1) & (mseccfg.RLB() == 0b0) & (v[7] == bitone) & (v[1] == bitone) | (mseccfg.MML() == 0b1) & (mseccfg.RLB() == 0b0) & (v[7] == bitone) & (v[1] == bitone) & (v[2] == bitone) | (mseccfg.MML() == 0b1) & (mseccfg.RLB() == 0b0) & (v[7] == bitone) & (v[0] == bitone) & (v[2] == bitone))
```
This is not readable

HamzaKh01 replied:
Condition is broken down for improved readability

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +162,28 @@ function pmpReadCfgReg(n) = {
+   then cfg
+   else {
+   if (mseccfg.MML() == 0b0) & (v[0] == bitzero) & (v[1] == bitone)
+   then Mk_Pmpcfg_ent(v & 0x9c)       // Bits 5 and 6 are zero and bits 0 and 1 are zero because of RW=01 reserver combination.
```
reserver?

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +162,28 @@ function pmpReadCfgReg(n) = {
 
 /* Helpers to handle locked entries */
-function pmpLocked(cfg: Pmpcfg_ent) -> bool =
+function pmpLockBit(cfg: Pmpcfg_ent) -> bool =
    cfg.L() == 0b1
+   
+function pmpLocked(cfg: Pmpcfg_ent) -> bool =
+   (cfg.L() == 0b1) & (mseccfg.RLB() == 0b0)
```
Maybe better to add a bool argument for whether to allow bypass... pmpcfg.L is still the lock bit and the entry is still locked if set, just in certain situations it can be bypassed when mseccfg.RLB is set

jrtc27 reviewed (commented):
Do you not also need 0x757 on RV32 for mseccfgh?

jrtc27 commented:
You have tabs in your changes

HamzaKh01 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +162,28 @@ function pmpReadCfgReg(n) = {
 function pmpWriteCfg(cfg: Pmpcfg_ent, v: bits(8)) -> Pmpcfg_ent =
-  if pmpLocked(cfg) then cfg
-  else Mk_Pmpcfg_ent(v & 0x9f)  // Bits 5 and 6 are zero.
+  if pmpLocked(cfg) then cfg   /*If locked then configuration is unchanged*/
```
Fixed

HamzaKh01 commented on model/riscv_pmp_regs.sail:
```diff
@@ -216,3 +237,15 @@ function pmpWriteAddr(locked: bool, tor_locked: bool, reg: xlenbits, v: xlenbits
+    then reg | (v & EXTZ(0b11))
+    else reg | (v & EXTZ(0b111))
+  }
+  else (reg & EXTZ(0b11)) | (v & EXTZ(0b111))  /*Implements stickiness if once set remains set*/
```
Added the new line

HamzaKh01 commented:
Should I remove tabs that are for indentation?

jrtc27 commented:
Yes, as documented in CODE_STYLE.md indentation is two spaces

HamzaKh01 commented:
Ok, I will remove them and what about other changes are they fine, and in addition "Do you not also need 0x757 on RV32 for mseccfgh?" I will come back to you about this question a little later.

marnovandermaas commented on model/riscv_pmp_control.sail:
```diff
@@ -92,6 +92,21 @@ function pmpAddrRange(cfg: Pmpcfg_ent, pmpaddr: xlenbits, prev_pmpaddr: xlenbits
 
+/* Check MMl=1 and mmwp=0 behaviour*/
+
+val MMLcheck: (AccessType(ext_access_type)) -> bool effect {rreg}
```
I would call this something else, it's a check for a region that is unmatched and MML is enabled. So maybe rename it `unmatchedMmlCheck`?

marnovandermaas commented on model/riscv_pmp_regs.sail:
```diff
@@ -216,3 +239,15 @@ function pmpWriteAddr(locked: bool, tor_locked: bool, reg: xlenbits, v: xlenbits
+    then reg | (v & EXTZ(0b11))
+    else reg | (v & EXTZ(0b111))
+  }
+  else (reg & EXTZ(0b11)) | (v & EXTZ(0b111))
```
Nit: missing enline

marnovandermaas commented on model/riscv_pmp_regs.sail:
```diff
@@ -163,30 +164,29 @@ function pmpReadCfgReg(n) = {
 /* Helpers to handle locked entries */
 function pmpLockBit(cfg: Pmpcfg_ent) -> bool =
-   cfg.L() == 0b1
+  cfg.L() == 0b1
```
For this commit can you make it just about adding 0x757 and split the formatting changes into a different commit to make it easier to review?

marnovandermaas commented on model/riscv_pmp_regs.sail:
```diff
@@ -163,30 +164,29 @@ function pmpReadCfgReg(n) = {
-      then Mk_Pmpcfg_ent(v & 0x9c)       /* Bits 5 and 6 are zero and Bits 0 and 1 are zero because of RW=01 reserved combination */
-    else Mk_Pmpcfg_ent(v & 0x9f)        /* Bits 5 and 6 are zero */
-   }
+	if ((mseccfg.MML() == 0b1) & (mseccfg.RLB() == 0b0) & (v[7] == bitone)) then {
+   if ((v[0] == bitzero & v[1] == bitzero & v[2] == bitone) |
+       (v[0] == bitzero & v[1] == bitone  & v[2] == bitzero)|
+       (v[0] == bitzero & v[1] == bitone  & v[2] == bitone) |
+       (v[0] == bitone  & v[1] == bitzero & v[2] == bitone))
+   then cfg                           /* If condition meets configuration remains same which means pmpcfg writes are ignored */
+   else Mk_Pmpcfg_ent(v & 0x9f)         /* Bits 5 and 6 are zero */
+  }
+  else {
+   if (mseccfg.MML() == 0b0) & (v[0] == bitzero) & (v[1] == bitone)
+   then Mk_Pmpcfg_ent(v & 0x9c)       /* Bits 5 and 6 are zero and Bits 0 and 1 are zero because of RW=01 reserved combination */
+   else Mk_Pmpcfg_ent(v & 0x9f)        /* Bits 5 and 6 are zero */
+  }
```
What has actually changed here? If it is just formatting, please put it in a separate formatting commit.

marnovandermaas reviewed (commented):
To keep a clean commit history can you remove adding the z3_problems files from the first commit instead of removing them in the second?

HamzaKh01 commented:
@marnovandermaas changes you are referring to regarding the MMLcheck function are fixed in a later commit as it was used just once so removed. In addition, the formatting changes are also fixed in a later commit.

@jrtc27 @marnovandermaas  To fix all problems related to formatting and functionality added that is 0x757 added later I am thinking of squashing two commits of formatting into my first commit and will separate the last commit(of 0x757 added) and its formatting too.

**As a result now as a final state, We will have only two commits**

1. The First one will contain support of 0x747 with formatting done by default now 
2. The second commit will now have the support of 0x757 on RV32 for mseccfgh

Please let me know so, I do that for clean commit history.

jrtc27 commented:
No, RV32 support should be part of the first commit as without it it’s not a complete implementation

HamzaKh01 commented:
@jrtc27  So, you mean everything in just one commit?? That will be a complete implementation. If yes, then let me know I will do that.

HamzaKh01 commented:
@jrtc27 Done.

jrtc27 commented on model/riscv_pmp_control.sail:
```diff
@@ -104,15 +104,46 @@ function pmpCheckRWX(ent, acc) = {
+        }
+      }
+    }
+    else {
+      if ent.R() == 0b1 & ent.W() == 0b1 & ent.X() == 0b1 & pmpLockBit(ent) then {
```
else if?.... That or fix indentation for the else below if it makes more sense to group these two as separate

HamzaKh01 replied:
Fixed

jrtc27 commented on model/riscv_pmp_control.sail:
```diff
@@ -218,8 +249,15 @@ function pmpCheck forall 'n, 'n > 0. (addr: xlenbits, width: atom('n), acc: Acce
     PMP_Continue => match priv {
-                      Machine => true,
-                      _       => false
+                      Machine => if mseccfg.MMWP() == 0b1             /* Make Read, Write and execute denied by default, if condition meets for M mode */
```
You have tabs all over the place in here and it renders as a total mess in GitHub

HamzaKh01 replied:
Removed tab space

jrtc27 commented on model/riscv_pmp_control.sail:
```diff
@@ -104,15 +104,46 @@ function pmpCheckRWX(ent, acc) = {
+      else false
+    }}
   }
+  else 
+    match priv {
+      Machine => if pmpLockBit(ent) then 
+                 pmpCheckRWX(ent, acc)
+                 else true,
+        _     => pmpCheckRWX(ent, acc)
+    }
```
You have trailing whitespace on some of these lines

HamzaKh01 replied:
Fixed

jrtc27 commented on model/riscv_pmp_control.sail:
```diff
@@ -218,8 +249,15 @@ function pmpCheck forall 'n, 'n > 0. (addr: xlenbits, width: atom('n), acc: Acce
+				                            Execute(_)  => false,
+      	                            	   _      => true
+			                            }}
+	                               else true,        
```
Trailing whitespace

HamzaKh01 replied:
Fixed

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +163,34 @@ function pmpReadCfgReg(n) = {
+/* Function to check whether executable region can be added or not in accordance to 4b of Smepmp spec */
+function checkExe(cfg: Pmpcfg_ent,v: bits(8)) -> Pmpcfg_ent =
+  if ((mseccfg.MML() == 0b1) & (mseccfg.RLB() == 0b0) & (v[7] == bitone)) then {
+   if ((v[0] == bitzero & v[1] == bitzero & v[2] == bitone) |
```
This has a one space indent?..

HamzaKh01 replied:
Removed this function and merged the check from where it was called

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +163,34 @@ function pmpReadCfgReg(n) = {
+function checkExe(cfg: Pmpcfg_ent,v: bits(8)) -> Pmpcfg_ent =
+  if ((mseccfg.MML() == 0b1) & (mseccfg.RLB() == 0b0) & (v[7] == bitone)) then {
+   if ((v[0] == bitzero & v[1] == bitzero & v[2] == bitone) |
+       (v[0] == bitzero & v[1] == bitone  & v[2] == bitzero)|
```
)| needs a space

HamzaKh01 replied:
Removed this function

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +163,34 @@ function pmpReadCfgReg(n) = {
+  (cfg.L() == 0b1) & (mseccfg.RLB() == 0b0) & (pmpAddrMatchType_of_bits(cfg.A()) == TOR)
+  
+/* Function to check whether executable region can be added or not in accordance to 4b of Smepmp spec */
+function checkExe(cfg: Pmpcfg_ent,v: bits(8)) -> Pmpcfg_ent =
```
`, `

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +163,34 @@ function pmpReadCfgReg(n) = {
+       (v[0] == bitzero & v[1] == bitone  & v[2] == bitzero)|
+       (v[0] == bitzero & v[1] == bitone  & v[2] == bitone) |
+       (v[0] == bitone  & v[1] == bitzero & v[2] == bitone))
+   then cfg                           /* If condition meets configuration remains same which means pmpcfg writes are ignored */
```
If you're going to try and line comments up, actually line them up

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +163,34 @@ function pmpReadCfgReg(n) = {
 /* Helpers to handle locked entries */
+function pmpLockBit(cfg: Pmpcfg_ent) -> bool =
+  cfg.L() == 0b1
+   
```
Trailing whitespace

HamzaKh01 replied:
Fixed

jrtc27 replied:
No it's not

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +163,34 @@ function pmpReadCfgReg(n) = {
-   (cfg.L() == 0b1) & (pmpAddrMatchType_of_bits(cfg.A()) == TOR)
-
+  (cfg.L() == 0b1) & (mseccfg.RLB() == 0b0) & (pmpAddrMatchType_of_bits(cfg.A()) == TOR)
+  
```
Trailing whitespace

HamzaKh01 replied:
Fixed by moving in one line

jrtc27 replied:
No it's not

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +163,34 @@ function pmpReadCfgReg(n) = {
+   then Mk_Pmpcfg_ent(v & 0x9c)       /* Bits 5 and 6 are zero and Bits 0 and 1 are zero because of RW=01 reserved combination */
+   else Mk_Pmpcfg_ent(v & 0x9f)        /* Bits 5 and 6 are zero */
+  }
+	
```
Trailing whitespace

HamzaKh01 replied:
Fixed

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -216,3 +244,15 @@ function pmpWriteAddr(locked: bool, tor_locked: bool, reg: xlenbits, v: xlenbits
+
+function mseccfgWrite(reg: xlenbits, v: xlenbits) -> xlenbits =   /* to set RLB, need to check PMPCFG_L */
+  if mseccfg.RLB() == 0b0 & v[2] == bitone then {
+   if (pmp0cfg.L() == 0b1 | pmp1cfg.L() == 0b1 | pmp2cfg.L() == 0b1 | pmp3cfg.L() == 0b1 
+      | pmp4cfg.L() == 0b1 | pmp5cfg.L() == 0b1 | pmp6cfg.L() == 0b1 | pmp7cfg.L() == 0b1 
+      | pmp8cfg.L() == 0b1 | pmp9cfg.L() == 0b1 | pmp10cfg.L() == 0b1 | pmp11cfg.L() == 0b1 
+      | pmp12cfg.L() == 0b1 | pmp13cfg.L() == 0b1 | pmp14cfg.L() == 0b1 | pmp15cfg.L() == 0b1) 
```
These all have trailing whitespace

HamzaKh01 replied:
Fixed

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +163,34 @@ function pmpReadCfgReg(n) = {
+/* Function to check whether executable region can be added or not in accordance to 4b of Smepmp spec */
+function checkExe(cfg: Pmpcfg_ent,v: bits(8)) -> Pmpcfg_ent =
+  if ((mseccfg.MML() == 0b1) & (mseccfg.RLB() == 0b0) & (v[7] == bitone)) then {
+   if ((v[0] == bitzero & v[1] == bitzero & v[2] == bitone) |
```
Please construct the new Pmpcfg_ent *first* and then legalise it based on its named fields rather than do all this opaque bit indexing

jrtc27 replied:
(See legalize_mstatus, for example)

HamzaKh01 replied:
Removed the opaque bit indexing by constructing 4 legalized Pmpcfg_ent and then made the comparison with the configuration to be written and made the decision based on that.

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +163,34 @@ function pmpReadCfgReg(n) = {
+  (cfg.L() == 0b1) & (mseccfg.RLB() == 0b0) & (pmpAddrMatchType_of_bits(cfg.A()) == TOR)
+  
+/* Function to check whether executable region can be added or not in accordance to 4b of Smepmp spec */
+function checkExe(cfg: Pmpcfg_ent,v: bits(8)) -> Pmpcfg_ent =
```
checkExe is a poor name for this

jrtc27 replied:
And I don't understand why this isn't just inlined in pmpWriteCfg, it's just one extra check

HamzaKh01 replied:
Yes right... removed the function

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -216,3 +244,15 @@ function pmpWriteAddr(locked: bool, tor_locked: bool, reg: xlenbits, v: xlenbits
   then { if (locked | tor_locked) then reg else v }
   else { if (locked | tor_locked) then reg else EXTZ(v[53..0]) }
+
+function mseccfgWrite(reg: xlenbits, v: xlenbits) -> xlenbits =   /* to set RLB, need to check PMPCFG_L */
```
Take the Mseccfg_ent?

HamzaKh01 replied:
Corrected the function based on this "Mseccfg_ent" now

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -216,3 +244,15 @@ function pmpWriteAddr(locked: bool, tor_locked: bool, reg: xlenbits, v: xlenbits
   else { if (locked | tor_locked) then reg else EXTZ(v[53..0]) }
+
+function mseccfgWrite(reg: xlenbits, v: xlenbits) -> xlenbits =   /* to set RLB, need to check PMPCFG_L */
+  if mseccfg.RLB() == 0b0 & v[2] == bitone then {
```
As with checkExe, minimise opaque bit vector indexing

HamzaKh01 replied:
it would hinder the readability I think this should be ok...as  simple to read

jrtc27 replied:
No it wouldn't. As it stands nobody reading this code has any clue what these magic sets of bits are. The whole point of the types is so that these magic bits get names that the code uses everywhere. You should only be using bitvectors when necessary, and never accessing bits within them directly.

testhamzak commented on model/riscv_pmp_control.sail:
```diff
@@ -104,15 +104,46 @@ function pmpCheckRWX(ent, acc) = {
+           _         => false
+        }
+      }
+      else {
```
resolved

HamzaKh01 replied:
Fixed

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +163,31 @@ function pmpReadCfgReg(n) = {
+function pmpTORLocked(cfg: Pmpcfg_ent) -> bool = (cfg.L() == 0b1) & (mseccfg.RLB() == 0b0) & (pmpAddrMatchType_of_bits(cfg.A()) == TOR)
+  
+function pmpWriteCfg(cfg: Pmpcfg_ent, v: bits(8)) -> Pmpcfg_ent = {
+  let RWX_001 : Pmpcfg_ent = Mk_Pmpcfg_ent(v[7 .. 3] @ 0b1 @ 0b0 @ 0b0);
+  let RWX_010 : Pmpcfg_ent = Mk_Pmpcfg_ent(v[7 .. 3] @ 0b0 @ 0b1 @ 0b0);
+  let RWX_011 : Pmpcfg_ent = Mk_Pmpcfg_ent(v[7 .. 3] @ 0b1 @ 0b1 @ 0b0);
+  let RWX_101 : Pmpcfg_ent = Mk_Pmpcfg_ent(v[7 .. 3] @ 0b1 @ 0b0 @ 0b1);
+  let val_to_write : Pmpcfg_ent = Mk_Pmpcfg_ent(v & 0xff);
+  if pmpLocked(cfg) then cfg                                          /* If locked then configuration is unchanged */
+  else {
+    if ((mseccfg.MML() == 0b1) & (mseccfg.RLB() == 0b0) & (v[7] == bitone)) then {
+      if ((val_to_write == RWX_001) | (val_to_write == RWX_010) | (val_to_write == RWX_011) | (val_to_write == RWX_101))
+      then cfg                                                        /* If condition meets configuration remains same which means pmpcfg writes are ignored */
+      else Mk_Pmpcfg_ent(v & 0x9f)                                    /* Bits 5 and 6 are zero */
+    } else {
+    if (mseccfg.MML() == 0b0) & (v[0] == bitzero) & (v[1] == bitone)
+    then Mk_Pmpcfg_ent(v & 0x9c)                                      /* Bits 5 and 6 are zero and Bits 0 and 1 are zero because of RW=01 reserved combination */
+    else Mk_Pmpcfg_ent(v & 0x9f)                                      /* Bits 5 and 6 are zero */
+    }
+  }
```
This is not what I said to do. Construct _one_ Pmpcfg_ent and use field accessors in your if statements. No bitvector indexing, and no. As for updating, use field updates, don't then go back to the bitvector, the whole point is to make the code self-documenting in terms of the fields it's touching.

HamzaKh01 replied:
**Fixed** ... The only concern your now be, line 173 of _riscv_pmp_regs.sail_ .But for constructing legal Pmpcfg_ent  vector/bit concatenation wasn't working which is why I did, like that otherwise everything is according to your comments. For reference, you can look into that snippet
![saill](https://user-images.githubusercontent.com/98050888/216312769-eca21686-e01d-4023-8180-a8556b8ee682.png)

HamzaKh01 commented:
> No it's not

Here you are referring to outdated file, in updated it is fixed moved in a single line

HamzaKh01 commented:
> No it wouldn't. As it stands nobody reading this code has any clue what these magic sets of bits are. The whole point of the types is so that these magic bits get names that the code uses everywhere. You should only be using bitvectors when necessary, and never accessing bits within them directly.

It is fixed completely according to comments

HamzaKh01 commented:
@jrtc27 kindly review this as soon as you get  time so, I can wind this up.

jrtc27 commented:
> > No it's not
> 
> Here you are referring to outdated file, in updated it is fixed moved in a single line

No I wasn't. I opened the up to date file in a text editor before commenting. You still have trailing whitespace, one instance of which is on the line associated with that comment.

HamzaKh01 commented:
> No it's not
>> Here you are referring to outdated file, in updated it is fixed moved in a single line
>>> No I wasn't. I opened the up to date file in a text editor before commenting. You still have trailing whitespace, one instance of which is on the line associated with that comment.

Fixed......... @jrtc27 Please see and let me know if everything is ok now !

HamzaKh01 commented:
Hi @jrtc27, if everything looks ok now, Can I mention Bill and Martin? because they asked me to inform them, once everything is resolved. Thank you

jrtc27 commented on model/riscv_pmp_control.sail:
```diff
@@ -104,15 +104,41 @@ function pmpCheckRWX(ent, acc) = {
+          _          => false
+        }
+      } else {
+      match acc {
```
*Please* read your own PR and fix all the indentation issues, many are not properly nested. It's tiring going through and saying the same thing over and over.

jrtc27 commented on model/riscv_pmp_control.sail:
```diff
@@ -218,9 +244,17 @@ function pmpCheck forall 'n, 'n > 0. (addr: xlenbits, width: atom('n), acc: Acce
                     }
+                 } else true,
+      _       => false                             /* Make Read, Write and execute denied by default for S/U mode */
+                   }
```
What on earth is this indentation

jrtc27 commented on model/riscv_pmp_control.sail:
```diff
@@ -218,9 +244,17 @@ function pmpCheck forall 'n, 'n > 0. (addr: xlenbits, width: atom('n), acc: Acce
+                 else if mseccfg.MML() == 0b1
+                 then {                            /* Make execute denied, if condition meets for M mode */
+                   match acc {
+                    Execute(_) => false,
```
Only 1 column indent?

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -100,6 +100,15 @@ bitfield Pmpcfg_ent : bits(8) = {
+}
+
+register mseccfg  : Mseccfg_ent
+register mseccfgh  : bits(32)
```
What's with the spaces? Line up or single space, don't do multiple spaces that don't line up.

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +163,30 @@ function pmpReadCfgReg(n) = {
+  else {
+    if (mseccfg.MML() == 0b1 & mseccfg.RLB() == 0b0 & legal_v.L() == 0b1) then {
+      if ((legal_v.R() == 0b0 & legal_v.W() == 0b0 & legal_v.X() == 0b1) |
+      (legal_v.R() == 0b0 & legal_v.W() == 0b1 & legal_v.X() == 0b0) |
```
Indent continuation lines

martinberger replied:
Why not use pattern matching. It's a subjective preference and I don't want to argue for my preference, but I find that complex nested conditionals are harder to read than pattern matching.

jrtc27 replied:
You're right, that could well make more sense in this instance actually

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -154,15 +163,30 @@ function pmpReadCfgReg(n) = {
+    else legal_v;
+  if pmpLocked(cfg) then cfg                                             /* If locked then configuration is unchanged */
+  else {
+    if (mseccfg.MML() == 0b1 & mseccfg.RLB() == 0b0 & legal_v.L() == 0b1) then {
```
else if?

jrtc27 replied:
Though this is really hard to read at the moment...

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -216,3 +240,23 @@ function pmpWriteAddr(locked: bool, tor_locked: bool, reg: xlenbits, v: xlenbits
+  let reg =
+    if reg.RLB() == 0b0 & legal_v.RLB() == 0b1 then {                    /* to set RLB, need to check PMPCFG_L */
+      if (pmp0cfg.L() == 0b1 | pmp1cfg.L() == 0b1 | pmp2cfg.L() == 0b1 | pmp3cfg.L() == 0b1
+      | pmp4cfg.L() == 0b1 | pmp5cfg.L() == 0b1 | pmp6cfg.L() == 0b1 | pmp7cfg.L() == 0b1
```
Indent

HamzaKh01 commented:
- Used **match** instead of if else at most places now


- Asked to add command line switch for epmp by Bill, added the switch, given table summarizes the behaviour of SAIL model


|Serial no.|   Tests ran from ACT   |command line switch pmp |command line switch epmp|  No. of tests passed | No. of tests failed|
|:---:|:---:               |:---:                   |:---:       |    :----:   |          :---: |
|1|   PMP tests  |      Off        |     Off                   |      7     | 15       | Here's this   |
|2|    PMP tests               |        On                |  Off   |22        | 0     |
|3|    PMP tests              |        Off                |  On   | 22        |  0     |
|4|    PMP tests               |        On                |  On   | 22        |  0    |
|5|    ePMP tests               |        Off                |  Off   | 8        |    63   |
|6|    ePMP tests               |        Off                |  On   | 71        |  0     |
|7|    ePMP tests               |        On                |  Off   |  ModelStuck       |
|8|    ePMP tests               |        On                |  On   | 71       |    0   |


### Above table can be read as

- As an example take entry 2, Ran PMP tests by turning on the PMP switch from command line and ePMP switch is off meanwhile, and all of the 22 tests are passing

### Entry 1, 5 and 7 explanation from the table

- If I turn off both the switches then some of the pmp tests are still passing the reason is these tests are checking maybe just the configurations set in pmpcfg csr, Switch does not prevent writing these csr it just prevent the effect of these written csr, if a test is checking maybe configuration is written properly in the csr then it is passing despite the switch being on or off (**That is the behaviour of model, regarding pmp switch, that is already present**)

- In entry 5 as **I added epmp switch** I infered the same behaviour of model that exist for pmp switch as explained in above bullet. Some tests are passing despite switch is off because they may be checking bits that are set in  mseccfg but the effect that these bits will cause is prevented because switch is off

- In entry 7 model is stuck because i am running epmp tests by turning on pmp switch and turning off epmp switch. PMP extension does not translate the pmpcfg csr as epmp extension does and tests are written to check epmp behaviour that is why model is stuck in a test (Note it is expected behaviour)

## Note
ePMP switch is superset of pmp switch as you can infer from the results in table. If ePMP switch is On pmp switch is redundant. Kindly guide me do I have interpreted the behaviour right??

Description of switch provided `I will add a command line switch,  "--Smepmp" that will enable the feature.  Default
behaviour is to NOT enable the feature.  I will export a function, "haveSmepmp()",
to the Sail model which will return true if the switch is set, false if it is not.  You will
need to enable/disable based on this function's return value.`           Used function  `haveSmepmp()` in line 69 and 70 of riscv_pmp_regs.sail compiled the model multiple times with true and false output to check behaviour written in above table. One confusion is Do i have to keep lines  69 and 70 of riscv_pmp_regs.sail because for the moment I do not have the function `haveSmepmp()`  it will not compile without it. because I have used that at numerous places

Both pmp and epmp tests are not merged in the ACT repo, they are for the time being just PR's

allenjbaum commented:
Strictly speaking, I think, this should be just like the D extension. IF D
extension is supported, and F-extension must also be supported.
I don't think there is a consistent story about how that is conveyed to
models though. E.g. if misa.F if off, then misa.D must be off, enforced by
hardware.

But there is no equivalent CSR for PMP or ePMP. The question is really -
from a command line perspective:
should  ePMP should imply PMP, or must we explicitly list it (and,
therefore, if it isn't listed, riscv-config should report an error.

I don't have strong opinions one way or the other, but I think having it
imply PMP is the way to go.
So it should not be possible to run tests with ePMP support without PMP
support,
and ePMP tests should NOT be run without explicit ePMP support
- the RVTEST_CASE macro conditions should restrict those tests from being
run in that case.

On Mon, Feb 13, 2023 at 5:15 AM HamzaKh01 ***@***.***> wrote:

>
>    -
>
>    Used *match* instead of if else at most places now
>    -
>
>    Asked to add command line switch for epmp by Bill, added the switch,
>    given table summarizes the behaviour of SAIL model
>
> Serial no. Tests ran from ACT command line switch pmp command line switch
> epmp No. of tests passed No. of tests failed
> 1 PMP tests Off Off 7 15
> 2 PMP tests On Off 22 0
> 3 PMP tests Off On 22 0
> 4 PMP tests On On 22 0
> 5 ePMP tests Off Off 8 63
> 6 ePMP tests Off On 71 0
> 7 ePMP tests On Off ModelStuck
> 8 ePMP tests On On 71 0 Above table can be read as
>
>    - As an example take entry 2, Ran PMP tests by turning on the PMP
>    switch from command line and ePMP switch is off meanwhile, and all of the
>    22 tests are passing
>
> Entry 1, 5 and 7 explanation from the table
>
>    -
>
>    If I turn off both the switches then some of the pmp tests are still
>    passing the reason is these tests are checking maybe just the
>    configurations set in pmpcfg csr, Switch does not prevent writing these csr
>    it just prevent the effect of these written csr, if a test is checking
>    maybe configuration is written properly in the csr then it is passing
>    despite the switch being on or off (*That is the behaviour of model,
>    regarding pmp switch, that is already present*)
>    -
>
>    In entry 5 as *I added epmp switch* I infered the same behaviour of
>    model that exist for pmp switch as explained in above bullet. Some tests
>    are passing despite switch is off because they may be checking bits that
>    are set in mseccfg but the effect that these bits will cause is prevented
>    because switch is off
>    -
>
>    In entry 7 model is stuck because i am running epmp tests by turning
>    on pmp switch and turning off epmp switch. PMP extension does not translate
>    the pmpcfg csr as epmp extension does and tests are written to check epmp
>    behaviour that is why model is stuck in a test (Note it is expected
>    behaviour)
>
> Note
>
> ePMP switch is superset of pmp switch as you can infer from the results in
> table. If ePMP switch is On pmp switch is redundant. Kindly guide me do I
> have interpreted the behaviour right??
>
> Description of switch provided I will add a command line switch,
> "--Smepmp" that will enable the feature. Default behaviour is to NOT enable
> the feature. I will export a function, "haveSmepmp()", to the Sail model
> which will return true if the switch is set, false if it is not. You will
> need to enable/disable based on this function's return value. Used
> function haveSmepmp() in line 69 and 70 of riscv_pmp_regs.sail compiled
> the model multiple times with true and false output to check behaviour
> written in above table. One confusion is Do i have to keep lines 69 and 70
> of riscv_pmp_regs.sail because for the moment I do not have the function
> haveSmepmp() it will not compile without it. because I have used that at
> numerous places
>
> Both pmp and epmp tests are not merged in the ACT repo, they are for the
> time being just PR's
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/sail-riscv/pull/196#issuecomment-1427927035>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHPXVJSVCIPIGUH7VUIIO6TWXIXYPANCNFSM6AAAAAAS7WSCWA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

billmcspadden-riscv commented:
Note from the discussion this morning:
The switches that I proposed (--Smepmp and --Zicond) will not be
implemented.
Rather,  we will follow the nomenclature the LLVM uses in the -march string.

The Sail functional interface will be the same as what I proposed.

Bill Mc.


On Tue, Feb 14, 2023 at 10:03 AM Allen Baum ***@***.***>
wrote:

> Strictly speaking, I think, this should be just like the D extension. IF D
> extension is supported, and F-extension must also be supported.
> I don't think there is a consistent story about how that is conveyed to
> models though. E.g. if misa.F if off, then misa.D must be off, enforced by
> hardware.
>
> But there is no equivalent CSR for PMP or ePMP. The question is really -
> from a command line perspective:
> should ePMP should imply PMP, or must we explicitly list it (and,
> therefore, if it isn't listed, riscv-config should report an error.
>
> I don't have strong opinions one way or the other, but I think having it
> imply PMP is the way to go.
> So it should not be possible to run tests with ePMP support without PMP
> support,
> and ePMP tests should NOT be run without explicit ePMP support
> - the RVTEST_CASE macro conditions should restrict those tests from being
> run in that case.
>
> On Mon, Feb 13, 2023 at 5:15 AM HamzaKh01 ***@***.***> wrote:
>
> >
> > -
> >
> > Used *match* instead of if else at most places now
> > -
> >
> > Asked to add command line switch for epmp by Bill, added the switch,
> > given table summarizes the behaviour of SAIL model
> >
> > Serial no. Tests ran from ACT command line switch pmp command line switch
> > epmp No. of tests passed No. of tests failed
> > 1 PMP tests Off Off 7 15
> > 2 PMP tests On Off 22 0
> > 3 PMP tests Off On 22 0
> > 4 PMP tests On On 22 0
> > 5 ePMP tests Off Off 8 63
> > 6 ePMP tests Off On 71 0
> > 7 ePMP tests On Off ModelStuck
> > 8 ePMP tests On On 71 0 Above table can be read as
> >
> > - As an example take entry 2, Ran PMP tests by turning on the PMP
> > switch from command line and ePMP switch is off meanwhile, and all of the
> > 22 tests are passing
> >
> > Entry 1, 5 and 7 explanation from the table
> >
> > -
> >
> > If I turn off both the switches then some of the pmp tests are still
> > passing the reason is these tests are checking maybe just the
> > configurations set in pmpcfg csr, Switch does not prevent writing these
> csr
> > it just prevent the effect of these written csr, if a test is checking
> > maybe configuration is written properly in the csr then it is passing
> > despite the switch being on or off (*That is the behaviour of model,
> > regarding pmp switch, that is already present*)
> > -
> >
> > In entry 5 as *I added epmp switch* I infered the same behaviour of
> > model that exist for pmp switch as explained in above bullet. Some tests
> > are passing despite switch is off because they may be checking bits that
> > are set in mseccfg but the effect that these bits will cause is prevented
> > because switch is off
> > -
> >
> > In entry 7 model is stuck because i am running epmp tests by turning
> > on pmp switch and turning off epmp switch. PMP extension does not
> translate
> > the pmpcfg csr as epmp extension does and tests are written to check epmp
> > behaviour that is why model is stuck in a test (Note it is expected
> > behaviour)
> >
> > Note
> >
> > ePMP switch is superset of pmp switch as you can infer from the results
> in
> > table. If ePMP switch is On pmp switch is redundant. Kindly guide me do I
> > have interpreted the behaviour right??
> >
> > Description of switch provided I will add a command line switch,
> > "--Smepmp" that will enable the feature. Default behaviour is to NOT
> enable
> > the feature. I will export a function, "haveSmepmp()", to the Sail model
> > which will return true if the switch is set, false if it is not. You will
> > need to enable/disable based on this function's return value. Used
> > function haveSmepmp() in line 69 and 70 of riscv_pmp_regs.sail compiled
> > the model multiple times with true and false output to check behaviour
> > written in above table. One confusion is Do i have to keep lines 69 and
> 70
> > of riscv_pmp_regs.sail because for the moment I do not have the function
> > haveSmepmp() it will not compile without it. because I have used that at
> > numerous places
> >
> > Both pmp and epmp tests are not merged in the ACT repo, they are for the
> > time being just PR's
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <https://github.com/riscv/sail-riscv/pull/196#issuecomment-1427927035>,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AHPXVJSVCIPIGUH7VUIIO6TWXIXYPANCNFSM6AAAAAAS7WSCWA
> >
> > .
> > You are receiving this because you are subscribed to this thread.Message
> > ID: ***@***.***>
> >
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/sail-riscv/pull/196#issuecomment-1429987129>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXROLOC76PXLZ57TAKDAFE3WXOUDNANCNFSM6AAAAAAS7WSCWA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Bill McSpadden
Formal Verification Engineer
RISC-V International
mobile:  503-807-9309

jrtc27 commented on model/riscv_mem.sail:
```diff
@@ -142,7 +142,7 @@ function checked_mem_read forall 'n, 0 < 'n <= max_mem_access . (t : AccessType(
 /* PMP checks if enabled */
 function pmp_mem_read forall 'n, 0 < 'n <= max_mem_access . (t : AccessType(ext_access_type), p : Privilege, paddr : xlenbits, width : atom('n), aq : bool, rl : bool, res: bool, meta : bool) -> MemoryOpResult((bits(8 * 'n), mem_meta)) =
-  if   (~ (plat_enable_pmp ()))
+  if   (~ ((plat_enable_pmp ()) | (haveSmepmp())))
```
Surely you can't have Smepmp without the base PMP? That is, surely haveSmepmp should imply plat_enable_pmp and thus be redundant?

HamzaKh01 replied:
Yes, you are right. I did know that this check is redundant waiting for this [PR](https://github.com/riscv/sail-riscv/pull/246) to get merged then will push the final changes.

jrtc27 commented on model/riscv_pmp_control.sail:
```diff
@@ -104,14 +104,41 @@ function pmpCheckRWX(ent, acc) = {
+    0b0 =>
+      match priv {
+        Machine => if pmpLockBit(ent)
+        then pmpCheckRWX(ent, acc)
```
Indentation

jrtc27 commented on model/riscv_pmp_regs.sail:
```diff
@@ -66,6 +66,9 @@
 /*=======================================================================================*/
 
+function haveSmepmp() -> bool =
+  return false
```
No need to return, and can just be a single line

jrtc27 replied:
Also why isn't this with the other haveFoo functions?

HamzaKh01 replied:
This is added temporarily and will not be the part of the final implementation. Just for checking purposes.

allenjbaum commented:
This:
    Surely you can't have Smepmp without the base PMP? That is, surely
haveSmepmp should imply plat_enable_pmp and thus be redundant?
is an interesting point - does Smepmp require PMP, or does Smepmp imply
PMP? - from the point of view of Sail.

Another way of looking at this is: should Sail assume that configurations
passed in are legal, or must it check they are legal?
havePMP and have_SmePMP are passed into Sail - should Sail assume that
the combination have_SmePMP & !have_PMP will never passed in?
If so, then perhaps the answer is that there is a separate checking stage,
and when you get to this code, it is known that illegal combination can't
happen.
The next question is whether this separate checking stage is in Sail or
separate from it (e.g. the riscv-config tool that was designed to
explicitly check for this)
That would mean that Sail can use imply rather than require, which I think
is what Krste has stated in other discussions,
e.g. RV32IMAFD is redundant, because RV32IMAD implies the F.
How is this handled for F and D extensions in Sail?


On Tue, Apr 25, 2023 at 12:26 AM Jessica Clarke ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In model/riscv_mem.sail
> <https://github.com/riscv/sail-riscv/pull/196#discussion_r1176101689>:
>
> > @@ -142,7 +142,7 @@ function checked_mem_read forall 'n, 0 < 'n <= max_mem_access . (t : AccessType(
>
>  /* PMP checks if enabled */
>  function pmp_mem_read forall 'n, 0 < 'n <= max_mem_access . (t : AccessType(ext_access_type), p : Privilege, paddr : xlenbits, width : atom('n), aq : bool, rl : bool, res: bool, meta : bool) -> MemoryOpResult((bits(8 * 'n), mem_meta)) =
> -  if   (~ (plat_enable_pmp ()))
> +  if   (~ ((plat_enable_pmp ()) | (haveSmepmp())))
>
> Surely you can't have Smepmp without the base PMP? That is, surely
> haveSmepmp should imply plat_enable_pmp and thus be redundant?
> ------------------------------
>
> In model/riscv_pmp_control.sail
> <https://github.com/riscv/sail-riscv/pull/196#discussion_r1176102584>:
>
> > +            _        => false
> +          },
> +        (0b1, 0b1, 0b1, 0b1) =>
> +          match acc {
> +            Read(_) => true,
> +            _       => false
> +          },
> +        (_, _, _, _) =>
> +          if (priv == Machine) == pmpLockBit(ent)
> +          then pmpCheckRWX(ent, acc)
> +          else false
> +      },
> +    0b0 =>
> +      match priv {
> +        Machine => if pmpLockBit(ent)
> +        then pmpCheckRWX(ent, acc)
>
> Indentation
> ------------------------------
>
> In model/riscv_pmp_control.sail
> <https://github.com/riscv/sail-riscv/pull/196#discussion_r1176109049>:
>
> > @@ -217,10 +244,19 @@ function pmpCheck forall 'n, 'n > 0. (addr: xlenbits, width: atom('n), acc: Acce
>    match pmpMatchEntry(addr, width, acc, priv, pmp15cfg, pmpaddr15, pmpaddr14) {
>      PMP_Success  => true,
>      PMP_Fail     => false,
> -    PMP_Continue => match priv {
> -                      Machine => true,
> -                      _       => false
> -                    }
> +    PMP_Continue =>
> +      match priv {
> +        Machine =>
> +          if mseccfg.MMWP() == 0b1 & haveSmepmp()                /* Make Read, Write and execute denied by default, if condition meets for M mode */
>
> These would be clearer with explicit parentheses (even if in this case
> only one association makes sense, since 1 & x is always just x)
> ------------------------------
>
> In model/riscv_pmp_regs.sail
> <https://github.com/riscv/sail-riscv/pull/196#discussion_r1176110682>:
>
> > @@ -66,6 +66,9 @@
>  /*  SUCH DAMAGE.                                                                         */
>  /*=======================================================================================*/
>
> +function haveSmepmp() -> bool =
> +  return false
>
> No need to return, and can just be a single line
> ------------------------------
>
> In model/riscv_pmp_regs.sail
> <https://github.com/riscv/sail-riscv/pull/196#discussion_r1176110910>:
>
> > @@ -66,6 +66,9 @@
>  /*  SUCH DAMAGE.                                                                         */
>  /*=======================================================================================*/
>
> +function haveSmepmp() -> bool =
> +  return false
>
> Also why isn't this with the other haveFoo functions?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/sail-riscv/pull/196#pullrequestreview-1399274823>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHPXVJSVZ3C33GRO5CCY6ITXC54BBANCNFSM6AAAAAAS7WSCWA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

HamzaKh01 commented on model/riscv_pmp_control.sail:
```diff
@@ -217,10 +244,19 @@ function pmpCheck forall 'n, 'n > 0. (addr: xlenbits, width: atom('n), acc: Acce
+    PMP_Continue =>
+      match priv {
+        Machine =>
+          if mseccfg.MMWP() == 0b1 & haveSmepmp()                /* Make Read, Write and execute denied by default, if condition meets for M mode */
```
Okay

ptomsich commented:
I am confused by the history on this PR:
- The original author states "This is added temporarily and will not be the part of the final implementation. Just for checking purposes."
- Allen raises a point regarding dependencies and interactions with other extensions that has not been addressed.

Is this PR final or should it have been marked as DRAFT?

HamzaKh01 commented:
> This: Surely you can't have Smepmp without the base PMP? That is, surely haveSmepmp should imply plat_enable_pmp and thus be redundant? is an interesting point - does Smepmp require PMP, or does Smepmp imply PMP? - from the point of view of Sail. Another way of looking at this is: should Sail assume that configurations passed in are legal, or must it check they are legal? havePMP and have_SmePMP are passed into Sail - should Sail assume that the combination have_SmePMP & !have_PMP will never passed in? If so, then perhaps the answer is that there is a separate checking stage, and when you get to this code, it is known that illegal combination can't happen. The next question is whether this separate checking stage is in Sail or separate from it (e.g. the riscv-config tool that was designed to explicitly check for this) That would mean that Sail can use imply rather than require, which I think is what Krste has stated in other discussions, e.g. RV32IMAFD is redundant, because RV32IMAD implies the F. How is this handled for F and D extensions in Sail?

Surely all your concerns will not arise because we are not supporting the Smepmp switch at the moment now.  The switch implementation is taking too long and this PR seems to be stalled despite being functionally correct. The way it will work now is, if pmp is enabled and mseccfg(CSR of Smepmp extension) is used then PMP permissions will behave as ePMP.

HamzaKh01 commented:
> * The original author states "This is added temporarily and will not be the part of the final implementation. Just for checking purposes."

For your context this PR was originally prepared to add just support for Smepmp extension. Later I was asked to modify the implementation to  support as if it is controlled by the command line switch and I was just using API of that switch and compiling with true and false values to check my implementation not passing any value from command line and I was suppose to get this implementation of switch. That is why I said it is temporary thing because I have to define the variable to some value (True or False) otherwise build will fail. 

> Allen raises a point regarding dependencies and interactions with other extensions that has not been addressed.

 I have addressed that in an above comment.

>Is this PR final or should it have been marked as DRAFT?

Yes, it is final. The reason why I have taken back switch implementation from this is [here](https://github.com/riscv-admin/dev-partners/issues/12#issuecomment-1589446947) . Can you please review it or ask someone, if any feedback comes I will fix appropriately.

UmerShahidengr commented:
@billmcspadden-riscv are you merging this PR in today's meeting? If no, then what deliverables are still required from our side. By looking at the latest comments of @HamzaKh01, this PR has no dependencies with switch implementation so it can be merged right away.

UmerShahidengr commented:
Since this PR is the part of an SoW submitted at DevPartners and it is also getting tracked [here](https://github.com/riscv-admin/dev-partners/issues/10). @jjscheel and I are also interested in closing this PR soon as we have been waiting ro complete this work for last 7 months.

marnovandermaas commented on model/riscv_pmp_control.sail:
```diff
@@ -106,14 +106,42 @@ function pmpCheckRWX(ent, acc) = {
-    _       => pmpCheckRWX(ent, acc)
+  match mseccfg.MML() {
+    0b1 =>
+      match (ent.R(), ent.W(), ent.L(), ent.X()) {
```
Can we match the order of these with the one in the Smepmp specification?
LRWX

HamzaKh01 replied:
Sure.

marnovandermaas commented on model/riscv_pmp_control.sail:
```diff
@@ -106,14 +106,42 @@ function pmpCheckRWX(ent, acc) = {
+      match (ent.R(), ent.W(), ent.L(), ent.X()) {
+        (0b0, 0b1, 0b1, _) =>
+          match acc {
+            Execute(_) => true,
```
LRWX = 0000 means an inaccessible region right? So Execute should be false in that case?

marnovandermaas commented on model/riscv_pmp_control.sail:
```diff
@@ -106,14 +106,42 @@ function pmpCheckRWX(ent, acc) = {
+            Write(_) => priv == Machine | ent.X() == 0b1,
+            _        => false
+          },
+        (0b1, 0b1, 0b1, 0b1) =>
```
This is already caught in the first match statement right?

HamzaKh01 replied:
No, it is not caught earlier. In the first match R is 0.

marnovandermaas commented on model/riscv_pmp_control.sail:
```diff
@@ -106,14 +106,42 @@ function pmpCheckRWX(ent, acc) = {
+          },
+        (0b1, 0b1, 0b1, 0b1) =>
+          match acc {
+            Read(_) => true,
```
This should be an inaccessible region right?

HamzaKh01 replied:
No, It should not be an inaccessible region. Spec says

> The encoding pmpcfg.LRWX=1111 can be used for sharing data between M-mode and S/U mode,
where both modes only have read-only access to the region. The rule remains locked so that any
further modifications to its associated configuration or address registers are ignored until a PMP
reset, unless mseccfg.RLB is set.

My implementation is according to the spec.

marnovandermaas commented on model/riscv_pmp_control.sail:
```diff
@@ -106,14 +106,42 @@ function pmpCheckRWX(ent, acc) = {
+          },
+        (_, _, _, _) =>
+          if (priv == Machine) == pmpLockBit(ent)
+          then pmpCheckRWX(ent, acc)
```
I think this isn't true for LRWX=1010 where X is zero but it is a shared execute only region.

marnovandermaas replied:
It's also not true for LRWX=1011 where the region is readable in M mode.

HamzaKh01 replied:
This check is not intended to check LRWX=1010 and LRWX=1011 behaviour. Instead it is checked in the first match statement at line 115 and working as intended.

marnovandermaas commented on model/riscv_pmp_control.sail:
```diff
@@ -106,14 +106,42 @@ function pmpCheckRWX(ent, acc) = {
+    0b0 =>
+      match priv {
+        Machine =>
+          if pmpLockBit(ent)
```
Has this been renamed from `pmpLocked`?

marnovandermaas commented on model/riscv_pmp_regs.sail:
```diff
@@ -158,15 +167,31 @@ function pmpReadCfgReg(n) = {
 /* Helpers to handle locked entries */
-function pmpLocked(cfg: Pmpcfg_ent) -> bool =
-   cfg.L() == 0b1
+function pmpLockBit(cfg: Pmpcfg_ent) -> bool = cfg.L() == 0b1
```
Maybe split the renaming of these functions into a separate commit?

HamzaKh01 replied:
I needed that here in the Smepmp implementation. It is not simply renaming as explained in this [comment](https://github.com/riscv/sail-riscv/pull/196#discussion_r1280540596)

marnovandermaas commented on model/riscv_pmp_regs.sail:
```diff
@@ -158,15 +167,31 @@ function pmpReadCfgReg(n) = {
-function pmpWriteCfg(cfg: Pmpcfg_ent, v: bits(8)) -> Pmpcfg_ent =
-  if pmpLocked(cfg) then cfg
-  else Mk_Pmpcfg_ent(v & 0x9f)  // Bits 5 and 6 are zero.
+function pmpTORLocked(cfg: Pmpcfg_ent) -> bool = (cfg.L() == 0b1) & (mseccfg.RLB() == 0b0) & (pmpAddrMatchType_of_bits(cfg.A()) == TOR)
```
You could used the pmpLocked function in here right?

marnovandermaas commented on model/riscv_pmp_regs.sail:
```diff
@@ -158,15 +167,31 @@ function pmpReadCfgReg(n) = {
+function pmpTORLocked(cfg: Pmpcfg_ent) -> bool = (cfg.L() == 0b1) & (mseccfg.RLB() == 0b0) & (pmpAddrMatchType_of_bits(cfg.A()) == TOR)
+
+function pmpWriteCfg(cfg: Pmpcfg_ent, v: bits(8)) -> Pmpcfg_ent = {
+  let legal_v : Pmpcfg_ent = Mk_Pmpcfg_ent(v & 0x9f);                   /* Constructing legal Pmpcfg_ent by making bit 5 and 6 zero */
```
Maybe put the comment on a separate line?

HamzaKh01 replied:
ok

marnovandermaas commented on model/riscv_pmp_regs.sail:
```diff
@@ -158,15 +167,31 @@ function pmpReadCfgReg(n) = {
+    (0b0, 0b0, 0b1) => update_W(legal_v, 0b0),
+    (_, _, _)       => legal_v
+  };
+  if pmpLocked(cfg)                                                     /* If locked then configuration is unchanged */
```
Maybe put comment on a the line before?

marnovandermaas commented on model/riscv_pmp_regs.sail:
```diff
@@ -158,15 +167,31 @@ function pmpReadCfgReg(n) = {
+  };
+  if pmpLocked(cfg)                                                     /* If locked then configuration is unchanged */
+  then cfg
+  else if (mseccfg.MML() == 0b1 & mseccfg.RLB() == 0b0 & legal_v.L() == 0b1)
```
Add a comment to say that this is to prevent adding a rule with execution privileges if MML is enabled unless RLB is set.

HamzaKh01 replied:
Sure

marnovandermaas commented on model/riscv_pmp_regs.sail:
```diff
@@ -222,3 +247,24 @@ function pmpWriteAddr(locked: bool, tor_locked: bool, reg: xlenbits, v: xlenbits
+  let legal_v : Mseccfg_ent = Mk_Mseccfg_ent(EXTZ(v[2 .. 0]));
+  let reg : Mseccfg_ent = match (reg.RLB(), legal_v.RLB()) {                   /* to set RLB, need to check PMPCFG_L */
+    (0b0, 0b1) =>
+      if (pmp0cfg.L() == 0b1 | pmp1cfg.L() == 0b1 | pmp2cfg.L() == 0b1 | pmp3cfg.L() == 0b1
+         | pmp4cfg.L() == 0b1 | pmp5cfg.L() == 0b1 | pmp6cfg.L() == 0b1 | pmp7cfg.L() == 0b1
+         | pmp8cfg.L() == 0b1 | pmp9cfg.L() == 0b1 | pmp10cfg.L() == 0b1 | pmp11cfg.L() == 0b1
+         | pmp12cfg.L() == 0b1 | pmp13cfg.L() == 0b1 | pmp14cfg.L() == 0b1 | pmp15cfg.L() == 0b1)
```
Is this hardcoding that we only have 16 regions? Can't we have up to 64 PMP regions?

HamzaKh01 replied:
I added Smepmp support. I don't think it is part of this PR. Maybe you are right but what I feel is this is more like scope of pmp functionality. I just added an Smepmp check on top of this.

marnovandermaas replied:
I think writing to MSECCFG should be part of this PR, but we should not be hardcoding that we only have 16 PMP registers.

marnovandermaas reviewed (changes requested):
I added some comments throughout the code. I'm afraid that this PR is not ready for merging as it is. My main concern is that I believe that `pmpCheckPerms` is not correctly implemented for MML.

HamzaKh01 commented:
> I added some comments throughout the code. I'm afraid that this PR is not ready for merging as it is. My main concern is that I believe that pmpCheckPerms is not correctly implemented for MML.

I have replied to each comment separately and I believe pmpCheckPerms is correctly implemented for MML. I have noted your suggestions for adding comments in new line and adding some comments in code. Waiting for any other feedback from others will push the feedback fixes once review settles.

HamzaKh01 commented:
> I think writing to MSECCFG should be part of this PR, but we should not be hardcoding that we only have 16 PMP registers.

Yes, you are right, but currently  we only have 16 pmpcfg entries defined in the model. So, I am not hard-coding this to 16. Model has only the capability of adding up to 16 regions. If regions are increased up to 64 later only then this check can check 64 entries lock. I have still not understood your comment about how i avoid hard-coding up to 16 regions. The only solution which exists is to add support for up to 64 regions by increasing pmpcfg entries from 16 to 64 in the model as far as I know.

allenjbaum commented:
This is a case where we need to add a command line (or YAML) parameter that indicates whether the DUT has 0, 16, or 64 PMP entries. This is going to permeate the existing PMP code, so it should be done as a separate PR

HamzaKh01 commented:
Yes, agreed.

billmcspadden-riscv commented:
I've added the command line support for Smepmp.  You can find my code on the
branch, haveSmepmp_billmcspadden (https://github.com/riscv/sail-riscv/tree/haveSmepmp_billmcspadden),
in the sail-riscv repo.

The command line switch is:  --enable-Smepmp
The Sail functional interface is:   bool haveSmepmp()

I would suggest that you merge my code onto your branch and make
a unified PR that incorporates both my code and yours.

KotorinMinami commented:
@HamzaKh01 
So, it's been 5 months since this project and still no further progress? If it's possible flower, could you let me integrate these code correspondences and resolve the conflicts? My latest work may be related to the mseccfg register.

Timmmm commented on model/riscv_sys_control.sail:
```diff
@@ -100,6 +100,9 @@ function is_CSR_defined (csr : csreg, p : Privilege) -> bool =
     0x343 => p == Machine, // mtval
     0x344 => p == Machine, // mip
 
+    0x747 => p == Machine, // mseccfg
+    0x757 => p == Machine & (sizeof(xlen) == 32), // mseccfgh
+
```
I don't think that is strictly correct. You can have `mseccfg` without PMPs. Really we need a `sys_has_mseccfg()` flag or something like that, because it is optional. (As far as I remember anyway.)

marnovandermaas replied:
I think you're right. Either way this code needs to be altered before we can merge.

Timmmm commented on model/riscv_pmp_control.sail:
Hmm I think this is handled in `phys_access_check()` but maybe it makes more sense here.

marnovandermaas replied:
Ah, then I should probably be using that function in my work as well. Thanks! This can thus be marked as resolved with relation to this PR.

KotorinMinami commented:
Perhaps we can discuss further in #601
